src/modules/uORB/SubscriptionBlocking.hpp

```diff
@@ -0,0 +1,151 @@
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#pragma once
+
+#include "SubscriptionCallback.hpp"
+
+#include <containers/LockGuard.hpp>
+#include <px4_time.h>
+
+namespace uORB
+{
+
+// Subscription with callback that schedules a WorkItem
+template<typename T>
+class SubscriptionBlocking : public SubscriptionCallback
+{
+public:
+	/**
+	 * Constructor
+	 *
+	 * @param meta The uORB metadata (usually from the ORB_ID() macro) for the topic.
+	 * @param interval_us The requested maximum update interval in microseconds.
+	 * @param instance The instance for multi sub.
+	 */
+	SubscriptionBlocking(const orb_metadata *meta, uint32_t interval_us = 0, uint8_t instance = 0) :
+		SubscriptionCallback(meta, interval_us, instance)
+	{
+		pthread_mutexattr_t attr;
+
+		int ret_attr_init = pthread_mutexattr_init(&attr);
+
+		if (ret_attr_init != 0) {
+			PX4_ERR("pthread_mutexattr_init failed, status=%d", ret_attr_init);
+		}
+
+		int ret_mutexattr_settype = pthread_mutexattr_settype(&attr, PTHREAD_PRIO_NONE);
+
+		if (ret_mutexattr_settype != 0) {
+			PX4_ERR("pthread_mutexattr_settype failed, status=%d", ret_mutexattr_settype);
+		}
+
+		int ret_mutex_init = pthread_mutex_init(&_mutex, &attr);
+		{
+			PX4_ERR("pthread_mutex_init failed, status=%d", ret_mutex_init);
+		}
+	}
+
+	virtual ~SubscriptionBlocking()
+	{
+		pthread_mutex_destroy(&_mutex);
+		pthread_cond_destroy(&_cv);
+	}
+
+	void call() override
+	{
+		// signal immediately if no interval, otherwise only if interval has elapsed
+		if ((_interval_us == 0) || (hrt_elapsed_time(&_last_update) >= _interval_us)) {
+			pthread_cond_signal(&_cv);
+		}
+	}
+
+	/**
+	 * Copy the struct if updated.
+	 * @param data The data reference where the struct will be copied.
+	 * @param timeout_us The requested timeout in microseconds, or 0 to wait indefinitely.
+	 *
+	 * @return true only if topic was updated and copied successfully.
+	 */
+	bool updateBlocking(T &data, uint32_t timeout_us = 0)
+	{
+		if (!_registered) {
+			registerCallback();
+		}
+
+		if (updated()) {
+			// copy immediately if updated
+			return copy(&data);
+
+		} else {
+			// otherwise wait
+
+			LockGuard lg{_mutex};
+
+			if (timeout_us == 0) {
+				if (pthread_cond_wait(&_cv, &_mutex) == 0) {
+					return update(&data);
+				}
+
+			} else {
+				// otherwise wait with timeout based on interval
+				struct timespec ts;
+
+				if (px4_clock_gettime(CLOCK_REALTIME, &ts) != 0) {
+					return false;
+				}
+
+				// Calculate an absolute time in the future
+				uint64_t nsecs = ts.tv_nsec + (timeout_us * 1000 * 1000 * 1000);
+				static constexpr unsigned billion = (1000 * 1000 * 1000);
+				ts.tv_sec += (nsecs / billion);
+				nsecs -= (nsecs / billion) * billion;
+				ts.tv_nsec = nsecs;
```

Comment:
We already have this code a couple of times. Please re-use a function.

Reply:
Looks like we already have `abstime_to_ts` in hrt.